### ISSUE TITLE
Stress testing

### ISSUE THREAD

**Issue by Sarfraz-droid** [measuredco/puck#549]
Puck seems to lag on stress test. Are there any solutions to resolve such issue??

Example - On adding multiple lists. dnd starts lagging.

https://github.com/user-attachments/assets/be3b5a57-ee2e-4ae0-8cb9-5ad5042e5c7d

**Comment by Sarfraz-droid** [comment]
@chrisvxd Tried fixing this. here's a workaround I implemented. Used `react-virtualized-auto-sizer` and conditional rendering to optimize the component renders.


https://github.com/user-attachments/assets/9f64c93e-4594-42cf-8917-143f276920ea

**Comment by chrisvxd** [comment]
Sorry for delay @Sarfraz-droid. It's unusual to see Puck lagging that much, but it might be environment related.

Two related tickets:

1. #556 - this will gut almost all of the dnd functionality, so there's no point in looking at this until that's done
2. #644 - a catch-all for various rendering issues, which would likely cover this.

I'm going to close this in favour of #644.

**Comment by Sarfraz-droid** [comment]
Hey, I was working inside puck and I found the underlying issue. Main issue is. On every drag, the entire page rerenders. Because it's currently using a nested architecture. I shifted it to storing objects in a global store. And rendering only the changed elements using zustand & useSyncExternalStore hook.

Could be something to consider but would require an entire state overhaul

Performance got improved by a lot! I guess using profiler puck had around 1500 rerenders on drag. It got reduced by  150.
